### PR TITLE
MINOR: Logs warning message when user invoke producer#flush within callback

### PR DESCRIPTION
**Pull request by frankvicky**
As title.
Please refer to the discussion of #17946 for further details

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: docs/upgrade.html [hunk 1]
```diff
@@ -159,6 +159,9 @@ <h5><a id="upgrade_400_notable" href="#upgrade_400_notable">Notable changes in 4
                         </li>
                         <li>The deprecated <code>sendOffsetsToTransaction(Map&lt;TopicPartition, OffsetAndMetadata&gt;, String)</code> method has been removed from the Producer API.
                         </li>
+                        <li>The <code>flush</code> method now detects potential deadlocks and logs a warning when used inside a callback.
```
Comment (chia7712) [inline docs/upgrade.html]: this is unnecessary as it is just a log.

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this patch

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 2]
```diff
@@ -1219,11 +1219,19 @@ private void ensureValidRecordSize(int size) {
      * flush all buffered records before performing the commit. This ensures that all the {@link #send(ProducerRecord)}
      * calls made since the previous {@link #beginTransaction()} are completed before the commit.
      * </p>
+     * <p>
+     * <b>Important:</b> This method should not be used within the callback provided to
+     * {@link #send(ProducerRecord, Callback)}.Invoking <code>flush()</code> in this context will cause a deadlock.
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: nit: Space before "Invoking" please.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. Just one nit to fix.

File: docs/upgrade.html (hunk 1 above)
Comment (AndrewJSchofield) [reply docs/upgrade.html]: Personally, I tend the other way on this. If someone is now getting errors logged, this explains why. No big deal either way, I suppose. Just my 2 cents.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java (hunk 2 above)
Comment (frankvicky) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Thanks for pointing out 😺

**Review by AndrewJSchofield — APPROVED** [review]

**Pull request merged**